Some NewGRF variables concerning railtypes

Eddi-z opened the pull request:
Add: Var 6A, a clone of Var 4A for querying poweredness compared to an arbitrary railtype

Eddi-z committed:
Add: [NewGRF] Var 63, querying poweredness compared to an arbitrary railtype

Eddi-z commented:
NML usage: "var[0x6a, 8, 0xFF, ELRL]"

andythenorth commented:
Tested, worked as expected for me.

andythenorth commented:
The intent is to use this in, varaction 2 for e.g. CB36.  

Without this var, railtype labels have to be checked individually, which:

1.  is potentially a long list
2. requires vehicle grf to know all possible railtypes in advance, rather than relying on railtype grf to specify compatibility and powered-ness

Eddi-z commented:
Open questions (that i can't answer myself):

 * choice of variable number?
 * skip the (redundant, because same as var4A) reverse lookup?
 * are there railtype introspection variables needed other than poweredness?

frosch123 commented:
I second the "skip the redundant part".

This variable seems to be about comparing two railtypes.
Besided poweredness there is not much to compare (compatibility is useless here).
* Which one is better? Higher speed limit or something? Newer intro date?

In addition it may make sense to extend 4A with more attributes about the present railtype:
* Track has catenary?

Maybe - but not necessarily - related:
* Sometimes vehicles want to know about their speed limit, and whether it is defined by the vehicle, the railtype, the bridge, the timetable, stopping, ...
Currently there are some acceleration/deceleration flags, but an actual number to compare with "current_speed" may be more useful.

Eddi-z commented:
> In addition it may make sense to extend 4A with more attributes about the present railtype:
> 
>     * Track has catenary?
i don't think "track has catenary" is a property of a railtype that can be meaningfully extracted from the arbitrary NewGRF data.

> Maybe - but not necessarily - related:
> 
>     * Sometimes vehicles want to know about their speed limit, and whether it is defined by the vehicle, the railtype, the bridge, the timetable, stopping, ...
>       Currently there are some acceleration/deceleration flags, but an actual number to compare with "current_speed" may be more useful.
my previous idea about that was a vehicle state variable with values:
* standing
* accelerating
* cruising (at max speed)
* decelerating
* unloading loading

possibly also add a bitset of what the reason for "cruising" is?
* resulting acceleration force was (near) 0 (vehicle power exhausted)
* railtype limit
* bridge limit
* curve limit
* ...?

i feel like there was another ticket where that discussion would fit better

independent of that, adding railtype max speed to var4A should be possible, however

frosch123 commented:
Catenary is defined by RTF_CATENARY.

Eddi-z commented:
Ok, but if we open that can of worms, people will request to know about 3rd rail. and the ability to issue sparks from somewhere other than the top of the vehicle, and...

(and by "people" i mean myself)

andythenorth commented:
So...is it easier if I just continue checking 'ELRL' and ignore NuTracks and friends? o_O

Eddi-z committed:
Add: [NewGRF] Var 4A also returns "has catenary" flag

Eddi-z commented:
Removed the lookup bit, and renamed the var to 63, because there's not enough relation to var 4A left to warrant a similarity in name.
New NML usage: "var[0x63, 0, 0xFF, ELRL]"

Added "has catenary" flag and railtype speed limit to var 4A (doesn't have to be the speed limit currently affecting the train, so i don't know how useful that is)

glx22 commented on src/newgrf_engine.cpp:
```diff
@@ -607,9 +607,16 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
+
 			RailType rt = GetTileRailType(v->tile);
-			return (HasPowerOnRail(Train::From(v)->railtype, rt) ? 0x100 : 0) | GetReverseRailTypeTranslation(rt, object->ro.grffile);
+			RailtypeInfo *rti = &_railtypes[rt]
```
missing ;

LordAro reviewed (dismissed):
Looks fine, will wait for a GRF-God to approve & merge

andythenorth commented:
We'll also want some docs once this is merged. Such admin :)

PeterN commented on src/newgrf_engine.cpp:
```diff
@@ -693,6 +700,12 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
+		case 0x63: { // Powerdness relative to arbitrary railtype
+			if (v->type != VEH_TRAIN) return 0;
+			RailType rt = GetTileRailType(v->tile);
+			return HasPowerOnRail((RailType)parameter, rt);
```
Parameter will need translating from RTT index to railtype.

Eddi-z commented:
So, i have no idea if this is the correct way to make a translation between grf railtype and game railtype.

Also, i have added a variable that makes that translation and then a reverse translation, this might be useful for checking whether two "logical" railtypes are mapped to the same internal railtype. this might need some more extensive testing

PeterN commented:
Hmm, why does a NewGRF need to check mappings? It should be transparent to it.

Eddi-z commented:
Well, my use case would be a wagon that can run on a 15t, 18t, 20t or 22t railtype, with different capacities. but if the railtype GRF doesn't provide any distinguished weight class types (as per the [standardized railtype scheme](https://newgrf-specs.tt-wiki.net/wiki/Standardized_Railtype_Scheme)), the different variations could be omitted.

PeterN commented:
Hmm, that sounds like information to be queried at load-time, not a varaction chain.

Eddi-z commented:
yeah, we need something that can be used in action6/7/9/D. turns out i have totally forgotten how these things work.

PeterN commented:
Action 07 condition type 0D/0E can be used to check if a rail-type label exists. That's been there for 10 years.

Eddi-z commented:
That is not enough to detect this case

Eddi-z commented:
Slightly more elaborate answer, let's assume i have the following railtype table in my GRF (pseudocode):
```
rtt {
 rail1: [NAAN, RAIL],
 rail2: [NABN, RAIL],
}
```
now, NML will happily generate code to detect the presence of the railtypes with the existing checks, and fill in the fallback types as needed.

now the cases i want to detect here are:
1. neither railtype is defined, and the fallback RAIL is used
2. both railtypes are defined, and are different
3. both railtypes are defined, but are mapped to the same railtype via alternate label

case 3 is not detectable with the current methods

PeterN commented:
Hmm, okay, alternate labels are a new one on me. Added 7 years ago...

Eddi-z commented:
Ok, this is probably now completely crazy (and untested)

Eddi-z commented:
So how it is meant to work now is, assuming same railtype table as above, and the following action D:
```
-1 * -1 // ID, length
        0D // Action D
        00 // Result Parameter
        02 // Operation: Subtraction
        BF BF // New Var 3F: Check railtype mapping, +80 to indicate not-a-parameter
        00 01 // Which railtypes to check (Var 3F parameter)
        00 00 // leftover bytes from "data" (DWORD)
```
then the result in parameter 00 should be 0 if both are mapped to the same type (case 1 or 3 above) or non-0 if different (case 2 above), which then could be checked in a following action 6,7 or 9, as appropriate

PeterN commented on src/newgrf_engine.cpp:
```diff
@@ -693,6 +700,12 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
 			return ret;
 		}
 
+		case 0x63: { // Powerdness relative to arbitrary railtype
```
Poweredness

PeterN commented:
Have we got any test cases for all these changes?

Eddi-z commented:
Afaik @andythenorth had some test case for the "switch power between ELRL and RAIL" hybrid engine, which needs testing on more complex tracksets. but i haven't seen anything concrete otherwise.

PeterN commented on src/newgrf.cpp:
```diff
@@ -6054,7 +6054,7 @@ static uint32 GetParamVal(byte param, uint32 *cond_val)
 	/* First handle variable common with VarAction2 */
 	uint32 value;
-	if (GetGlobalVariable(param - 0x80, &value, _cur.grffile)) return value;
+	if (GetGlobalVariable(param - 0x80, &value, _cur.grffile, cond_val==NULL?0:*cond_val)) return value;
```
Spacing

LordAro commented:
Hi, there's been no activity on this for over 5 months. If no more effort will be put into this PR, it will be closed in the next week or so

Eddi-z commented:
I'm having 3 thoughts about this PR right now:

1. the speed limit stuff feels slightly too ambitious right now, and the raw railtype speed limit doesn't make much sense on its own, it might be better to leave it out right now
2. i need a second opinion on the global-var-with-parameter bit, as it might be a tiny step too hacky
3. some bits of this should maybe be extended for roadtypes, but that could be a separate PR

frosch123 commented:
1. Seems like I suggested a "maximum speed" variable. However, it seems that this is already available in var 0x4C. The pure "railtype speed limit" appears useless to me. So, remove.
2. Yes, this is completely different from what NewGRF usually does.
2a) VarAction2 variables with parameters must be in range 0x60-07F, otherwise the parameter is unusable.
2b) If the variable is not available in VA2, and it does not seem to make sense in ActionD either, then why assign a variable at all?
2c) Adding a new Action7 operator "check if railtypes are mapped to the same type" seems like the more natural way to do it.
2d) One could either pack two RTT-indices into a 2-byte sized value, or put two RT labels into a 8-byte sized value (as for var 0x88).
2e) Is there consensus on what to return if both railtypes do not exist?

frosch123 commented on src/newgrf_engine.cpp:
```diff
@@ -610,8 +610,14 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
 					RailType rt = GetTileRailType(v->tile);
-					return (HasPowerOnRail(Train::From(v)->railtype, rt) ? 0x100 : 0) | GetReverseRailTypeTranslation(rt, object->ro.grffile);
+					RailtypeInfo *rti = &_railtypes[rt];
+					return (rti->max_speed << 16) |
```
Please remove max_speed, as commented above.

frosch123 commented on src/newgrf_engine.cpp:
```diff
@@ -610,8 +610,14 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
-					return (HasPowerOnRail(Train::From(v)->railtype, rt) ? 0x100 : 0) | GetReverseRailTypeTranslation(rt, object->ro.grffile);
+					RailtypeInfo *rti = &_railtypes[rt];
+					return (rti->max_speed << 16) |
+						(HasPowerOnRail(Train::From(v)->railtype, rt) ? 0x100 : 0) |
```
Please order the bitfields big-endian: 0x200, 0x100, rt.

frosch123 commented on src/newgrf_engine.cpp:
```diff
@@ -704,6 +710,12 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
+		case 0x63: { // Poweredness relative to arbitrary railtype
+			if (v->type != VEH_TRAIN) return 0;
+			RailType rt = GetTileRailType(v->tile);
+			return HasPowerOnRail(object->ro.grffile->railtype_map[parameter], rt);
```
Use ? : to define an explicit value, instead of using implicit bool->int conversion.

Eddi-z committed:
Add: [NewGRF] Action 7 condition for checking whether two railtypes are identical/different

Eddi-z commented:
>     1.  So, remove.

Done
>        2c) Adding a new Action7 operator "check if railtypes are mapped to the same type" seems like the more natural way to do it.

added this now:
- 0x13: check whether two railtypes are the same
- 0x14: check whether two railtypes are different

>        2d) One could either pack two RTT-indices into a 2-byte sized value, or put two RT labels into a 8-byte sized value (as for var 0x88).

went with a 2-byte sized value, as the 8-byte stuff seems too weird/hacked to me, which would put it back into the same problematic category as the var3F stuff

>        2e) Is there consensus on what to return if both railtypes do not exist?

went with "check will fail if either railtype is not defined"

LordAro commented on src/newgrf.cpp:
```diff
@@ -6659,6 +6666,18 @@ static void SkipIf(ByteReader *buf)
+				}
+			case 0x14: {
+					RailType rt1 = GetRailTypeByLabel(_cur.grffile->railtype_list[GB(cond_val, 0, 8)]);
+					RailType rt1 = GetRailTypeByLabel(_cur.grffile->railtype_list[GB(cond_val, 8, 8)]);
```
copy & paste fail here

LordAro commented on src/newgrf.cpp:
```diff
@@ -6659,6 +6666,18 @@ static void SkipIf(ByteReader *buf)
 			case 0x12: result = GetRoadTypeByLabel(BSWAP32(cond_val)) != INVALID_ROADTYPE;
 				break;
+			case 0x13: {
+					RailType rt1 = GetRailTypeByLabel(_cur.grffile->railtype_list[GB(cond_val, 0, 8)]);
```
double indent for both these cases

Eddi-z replied:
Is that somewhere in the code style? this way felt the most natural to me, otherwise the closing `}` would be at the same indentation level as the next `case`

LordAro replied:
https://wiki.openttd.org/Coding_style#Control_flow yup :)

LordAro commented on src/newgrf_engine.cpp:
```diff
@@ -610,8 +610,13 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
 					RailType rt = GetTileRailType(v->tile);
-					return (HasPowerOnRail(Train::From(v)->railtype, rt) ? 0x100 : 0) | GetReverseRailTypeTranslation(rt, object->ro.grffile);
+					RailtypeInfo *rti = &_railtypes[rt];
+					return ((rti->flags & RTFB_CATENARY) ? 0x200 : 0) |
```
tempted to suggest these are complicated enough to be split into separate variables, to make it clearer which values go where. Nesting terneries inside bit-math is a recipe for confusion later...

Eddi-z replied:
I'm not quite following this argument.

LordAro replied:
as in have variables like `catenary_flag` & `power_flag` factored out before or-ing them together `catenary_flag | power_flag | ...`
Would look clearer, imo

frosch123 replied:
I prefer the ? : in place, like it currently is in this PR, and like it is for most variables and also for p1 and p2 in commands.

Usually you read this code in reverse, from the NewGRF point of view. So you want to know where does 0x100 come from. Distributing this information over multiple expressions makes it harder to read.

LordAro replied:
Fair enough, I wasn't looking at it in the wider context

LordAro commented on src/newgrf.cpp:
```diff
@@ -6252,7 +6253,7 @@ static void SkipAct5(ByteReader *buf)
  * @return true iff the variable is known and the value is returned in 'value'.
  */
-bool GetGlobalVariable(byte param, uint32 *value, const GRFFile *grffile)
+bool GetGlobalVariable(byte param, uint32 *value, const GRFFile *grffile, uint32 parameter)
```
parameter should be documented

Eddi-z replied:
I'll probably remove this commit

LordAro replied:
Still a thing...

andythenorth commented:
Example use case https://www.tt-forums.net/viewtopic.php?f=26&t=68616&start=520#p1227070

nielsmh commented on src/newgrf_engine.cpp:
```diff
@@ -704,6 +704,12 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
 		}
 
+		case 0x63: { // Poweredness relative to arbitrary railtype
+			if (v->type != VEH_TRAIN) return 0;
```
Should this be extended for NRT? Roadtypes can also be powered as far as I know.

michicc replied:
Road/tramtypes don't distinguish powered and compatible, though.

frosch123 replied:
IIRC the use-case for this is: A vehicle is dual-powered, it is powered on normal rail, but on electrified rail it raises a pantograph.
The same use-case is valid for roadtypes.

Eddi-z replied:
i now added the road-catenary flag to var 0x4A, however, i don't see how this would apply to var 0x63, other than returning "1" all the time.

frosch123 commented:
Last comment from a NewGRF author is over a year ago.
So there does not seem to be any demand for this.

Eddi-z commented:
uhm wtf kind of reasoning is that?!?

EmperorJake commented:
As a NewGRF Author, I demand this

michicc commented:
Needs a rebase. And I do think that it can improve handling of dual/multi-mode engines for NewGRFs.

andythenorth commented:
This could have been a simple single-issue change a long time ago, and I'd have used it.  

But I know that stacking single-issue changes is how we get an incoherent newgrf spec.

Stability vs. innovation eh.

Eddi-z commented:
rebased, and removed the Var 3F part that was probably too hacky

michicc commented on src/newgrf_engine.cpp:
```diff
@@ -702,6 +707,12 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
+		case 0x63: { // Poweredness relative to arbitrary railtype
+			if (v->type != VEH_TRAIN) return 0;
+			RailType rt = GetTileRailType(v->tile);
+			return HasPowerOnRail(object->ro.grffile->railtype_map[parameter], rt) ? 1 : 0;
```
Railtype map has as length of RAILTYPE_END (64 currently), while a buggy NewGRF could pass any parameter. I.e. missing limit check.

Eddi-z commented:
also added range checks for Action 7

michicc commented:
Principal approval from me.

frosch123 commented on src/newgrf_engine.cpp:
```diff
@@ -608,8 +608,13 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
+
 					RailType rt = GetTileRailType(v->tile);
-					return (HasPowerOnRail(Train::From(v)->railtype, rt) ? 0x100 : 0) | GetReverseRailTypeTranslation(rt, object->ro.grffile);
+					RailtypeInfo *rti = &_railtypes[rt];
```
could be "const"

frosch123 commented on src/newgrf.cpp:
```diff
@@ -6589,6 +6589,31 @@ static void SkipIf(ByteReader *buf)
+				}
+				RailType rt1 = GetRailTypeByLabel(_cur.grffile->railtype_list[r1]);
+				RailType rt2 = GetRailTypeByLabel(_cur.grffile->railtype_list[r2]);
+				result = rt1 != INVALID_RAILTYPE && rt2 != INVALID_RAILTYPE && rt1 != rt2;
```
This is not the negation of condition 0x13, which makes it weird in corner cases, and hard to implement in NML.

Eddi-z replied:
that is intentional. in case one of the railtypes is undefined, both conditions return "false"

Eddi-z replied:
maybe i'm too deep in formal logic here, but comparing `<undefined>` to `<undefined>` should never return true

Eddi-z replied:
... and since we can't return `<undefined>`, the only option left is to return false

frosch123 commented on src/newgrf.cpp:
```diff
@@ -6589,6 +6589,31 @@ static void SkipIf(ByteReader *buf)
+				uint r1 = GB(cond_val, 0, 8);
+				uint r2 = GB(cond_val, 8, 8);
+				if (r1 >= _cur.grffile->railtype_list.size() || r2 >= _cur.grffile->railtype_list.size()) {
+					result = false;
```
Similar places issue a grfmsg() in this case. Though not sure whether this case always triggers in GRF loading stages before reservation.

Eddi-z replied:
i'll take suggestions on what debuglevel this should be shown.

as for earlier loading stages, i guess that applies to action 9 only? not sure if the check makes any sense there.

michicc commented:
For some reason I can't reply to the first comment from frosch:

Roadtypes don't distinguish compatible and powered. If an RV can drive on it, it is by definition also powered and vice versa.

Eddi-z commented:
> Roadtypes don't distinguish compatible and powered. If an RV can drive on it, it is by definition also powered and vice versa.

i kinda see his point though. historically, there have been busses which can run on diesel or electric (catenary) mode, and maybe should make that visible.

michicc commented:
@Eddi-z After further though, yes. So that var should preferably also work for RVs.

Eddi-z commented on src/newgrf_engine.cpp:
```diff
@@ -702,6 +712,13 @@ static uint32 VehicleGetVariable(Vehicle *v, const VehicleScopeResolver *object,
+			if (v->type != VEH_TRAIN) return 0;
+			if (parameter >= RAILTYPE_END) return 0;
+			RailType rt = GetTileRailType(v->tile);
+			return HasPowerOnRail(object->ro.grffile->railtype_map[parameter], rt) ? 1 : 0;
```
i'm suddenly unsure whether i should have used railtype_list instead here, like in the action7 changes

frosch123 commented:
I created a new PR #8554 to implement the ideas from this PR.
I salvaged a bit of the code of this PR, but most was unusable/broken.

Please put more effort into thinking what a PR should do, and what cases must be considered, before coding and submitting. This helps missing 90% of the things to consider.
